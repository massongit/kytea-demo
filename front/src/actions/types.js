/**
 * KyTeaによる解析結果の表示Actionを表すタイプ
 * @type {string}
 */
export const SHOW_SENTENCE = "SHOW_SENTENCE"

/**
 * 品詞や読みの表示Actionを表すタイプ
 * @type {string}
 */
export const SHOW_PRONUNCIATION = "SHOW_PRONUNCIATION"

/**
 * ローディングActionを表すタイプ
 * @type {string}
 */
export const LOADING = "LOADING"
