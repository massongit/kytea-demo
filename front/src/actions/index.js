import * as types from "./types"
import {createAction} from "redux-actions"
import {intl} from "../intl"

/**
 * KyTeaによる解析結果の表示ActionのActionCreator
 * @type {function}
 */
export const showSentence = createAction(types.SHOW_SENTENCE, parameter => {
    // KyTeaによる解析結果の表示Action
    const action = {
        words: []
    }

    // 空白を除去したsentenceをActionにセット
    if (parameter.sentence) {
        action.sentence = parameter.sentence.replace(/\s/g, "")
    }

    if (parameter.words) {
        // 予測不能であった場合のタグ
        const unknownTag = "(Unknown)"

        for (const word of parameter.words) {
            if (word) {
                // Actionのword
                const word_ = {
                    word: word.word,
                    pronunciation: []
                }

                if (word.pos === unknownTag) {
                    word.pos = intl.formatMessage({
                        id: "unknown"
                    })
                } else {
                    word_.pos = word.pos
                }

                for (let i = 0; word.pronunciation && i < word.pronunciation.length; i++) {
                    if (word.pronunciation[i].pronunciation === unknownTag) {
                        word.pronunciation[i].pronunciation = intl.formatMessage({
                            id: "unknown"
                        })
                    }

                    word_.pronunciation.push(word.pronunciation[i])
                }

                action.words.push(word_)
            } else {
                action.words.push(word)
            }
        }
    }

    return action
})

/**
 * 品詞や読みの表示ActionのActionCreator
 * @type {function}
 */
export const showPOSAndPronunciation = createAction(types.SHOW_PRONUNCIATION, parameter => ({
    number: parameter.number,
    word: parameter.word,
    pos: parameter.pos,
    pronunciation: parameter.pronunciation
}))
