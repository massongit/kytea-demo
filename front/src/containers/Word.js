import Word from "../components/Word"
import {connect} from "react-redux"
import {injectIntl} from "react-intl"

/**
 * Sentenceの各ボタンのContainer
 */
export default injectIntl(connect()(Word))
