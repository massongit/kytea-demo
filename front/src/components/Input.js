import React from "react"
import fetch from "node-fetch"
import PropTypes from "prop-types"
import {Button, Form, FormControl, InputGroup} from "react-bootstrap"
import {showSentence} from "../actions"

/**
 * 入力部
 */
class Input extends React.Component {
    render() {
        return (
            <Form
                onSubmit={async ev => {
                    // サーバーへのSubmitが行われないようにする
                    ev.preventDefault()

                    this.input.value = this.input.value.trim()

                    // 以前の入力文とは異なる文章が入力されたとき、KyTeaによる解析を行い、KyTeaによる解析結果の表示Actionをdispatchする
                    if (0 < this.input.value.length && this.input.value !== this.props.sentence) {
                        try {
                            this.props.dispatch(showSentence({
                                sentence: this.input.value,
                                words: await (await fetch(
                                    "/kytea",
                                    {
                                        method: "POST",
                                        body: this.input.value
                                    }
                                ).then(res => {
                                    if (res.ok) {
                                        return res
                                    } else {
                                        throw new Error(res.statusText)
                                    }
                                })).json()
                            }))
                        } catch (er) {
                            alert("予測中にエラーが発生しました！\n" + er.message)
                        }
                    }
                }}
            >
                <InputGroup>
                    <FormControl
                        inputRef={n => {
                            this.input = n
                        }}
                    />
                    <InputGroup.Button>
                        <Button type="submit">
                            予測
                        </Button>
                    </InputGroup.Button>
                </InputGroup>
            </Form>
        )
    }
}

Input.propTypes = {
    dispatch: PropTypes.func.isRequired,
    sentence: PropTypes.string
}

export default Input
