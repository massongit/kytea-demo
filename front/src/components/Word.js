import React from "react"
import PropTypes from "prop-types"
import {Button, ButtonDropdown, DropdownItem, DropdownMenu, DropdownToggle} from "reactstrap"
import {intlShape} from "react-intl"
import {propTypesPart} from "./App"
import {showPOSAndPronunciation} from "../actions"

/**
 * Sentenceの各ボタン
 */
class Word extends React.Component {
    static propTypes = {
        dispatch: PropTypes.func.isRequired,
        word: PropTypes.string,
        intl: intlShape.isRequired,
        ...propTypesPart
    }

    constructor(props) {
        super(props)
        this.toggleDropdown = this.toggleDropdown.bind(this)
        this.state = {
            dropdownOpen: false
        }
    }

    /**
     * onClickイベント
     * @param n {number} 候補No (0起点)
     */
    onClick(n) {
        // 品詞や読みの表示Actionをdispatchする
        this.props.dispatch(showPOSAndPronunciation({
            number: n + 1,
            word: this.props.word,
            pos: this.props.pos,
            pronunciation: this.props.pronunciation[n]
        }))
    }

    /**
     * ボタン内のテキストを取得する
     * @param text {string} ボタン内のテキスト
     * @param n {number} 候補No (0起点)
     * @returns {string|element} ボタン内のテキスト
     */
    getButtonText(text, n) {
        if (this.props.pronunciation[n] && this.props.pronunciation[n].id) { // 読みが予測されているとき、文字列を強調
            return (
                <strong>
                    <u>
                        {text}
                    </u>
                </strong>
            )
        } else {
            return text
        }
    }

    /**
     * Dropdownの状態を切り替える
     */
    toggleDropdown() {
        this.setState(state => ({
            dropdownOpen: !state.dropdownOpen
        }))
    }

    /**
     * DropdownItemのリストを生成する
     * @returns {Array} DropdownItemのリスト
     */
    generateDropDownItems() {
        const dropdownItems = []

        for (let i = 0; i < this.props.pronunciation.length; i++) {
            // 今見ている読みが0番目の読みでないとき、区切り線をリストに追加
            if (0 < i) {
                dropdownItems.push(<DropdownItem divider/>)
            }

            // DropdownItemをリストに追加
            dropdownItems.push(this.renderDropdownItem(i))
        }

        return dropdownItems
    }

    /**
     * Buttonを描画する
     * @returns {node} Button
     */
    renderButton() {
        return (
            <Button
                outline
                color="dark"
                onClick={() => {
                    this.onClick(0)
                }}
            >
                {this.getButtonText(this.props.word, 0)}
            </Button>
        )
    }

    /**
     * DropdownButtonを描画する
     * @returns {node} DropdownButton
     */
    renderDropdownButton() {
        return (
            <ButtonDropdown
                isOpen={this.state.dropdownOpen}
                toggle={this.toggleDropdown}
            >
                <DropdownToggle
                    caret
                    color="outline-dark"
                >
                    {this.props.word}
                </DropdownToggle>
                <DropdownMenu>
                    {this.generateDropDownItems()}
                </DropdownMenu>
            </ButtonDropdown>
        )
    }

    /**
     * DropdownItemを描画する
     * @param n 候補No
     * @returns {node} DropdownItem
     */
    renderDropdownItem(n) {
        return (
            <DropdownItem
                key={n}
                onClick={() => {
                    this.onClick(n)
                }}
            >
                {
                    this.getButtonText(this.props.intl.formatMessage(
                        {
                            id: "candidate"
                        },
                        {
                            no: this.props.intl.formatNumber(n + 1)
                        }
                    ), n)
                }
            </DropdownItem>
        )
    }

    render() {
        if (this.props.pronunciation.length < 2) { // 読みの候補が複数存在しないとき、Buttonを出力
            return this.renderButton()
        } else { // 読みの候補が複数存在するとき、DropdownButtonを出力
            return this.renderDropdownButton()
        }
    }
}

export default Word
