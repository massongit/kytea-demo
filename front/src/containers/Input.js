import Input from "../components/Input"
import {connect} from "react-redux"
import {injectIntl} from "react-intl"

const mapStateToProps = state => ({
    sentence: state.showSentence.sentence,
    loading: state.loading.loading
})

/**
 * 入力部のContainer
 */
export default injectIntl(connect(mapStateToProps)(Input))
