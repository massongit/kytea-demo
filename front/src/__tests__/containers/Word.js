import React from "react"
import Word from "../../containers/Word"
import thunk from "redux-thunk"
import configureMockStore from "redux-mock-store"
import {shallow} from "enzyme"
import {eventMock} from "./Input"
import {showPOSAndPronunciation} from "../../actions"
import {
    initialShowSentenceState,
    initialShowPOSAndPronunciationState,
    pos,
    pronunciation_,
    showPOSAndPronunciationState_,
    word
} from "../reducers"

let wordComponent, store

describe("containers/Word", () => {
    beforeEach(() => {
        store = configureMockStore([thunk])({
            showSentence: initialShowSentenceState,
            showPOSAndPronunciation: initialShowPOSAndPronunciationState
        })
        wordComponent = shallow(
            <Word
                store={store}
                word={word}
                pos={pos}
                pronunciation={[pronunciation_]}
            />
        ).dive()
    })

    it("Componentが正しく配置されている", () => {
        expect(wordComponent).toMatchSnapshot()
    })

    it("wordが記述されている", () => {
        expect(wordComponent.children().contains(word)).toBeTruthy()
    })

    it("onClickイベントが呼び出されたとき、正常にdispatchが行われる", () => {
        wordComponent.simulate("click", eventMock)
        expect(store.getActions()).toEqual([
            showPOSAndPronunciation(showPOSAndPronunciationState_)
        ])
    })
})
