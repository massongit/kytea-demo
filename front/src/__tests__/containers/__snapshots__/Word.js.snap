// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Input Componentが正しく配置されている 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Input
    dispatch={[Function]}
    sentence=""
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
    storeSubscription={
        Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "bsClass": "form",
      "children": <InputGroup
        bsClass="input-group"
>
        <FormControl
                bsClass="form-control"
                componentClass="input"
                inputRef={[Function]}
        />
        <InputGroupButton
                bsClass="input-group-btn"
        >
                <Button
                        active={false}
                        block={false}
                        bsClass="btn"
                        bsStyle="default"
                        disabled={false}
                        type="submit"
                >
                        <FormattedMessage
                                id="predict"
                                values={Object {}}
                        />
                </Button>
        </InputGroupButton>
</InputGroup>,
      "componentClass": "form",
      "horizontal": false,
      "inline": false,
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bsClass": "input-group",
        "children": Array [
          <FormControl
            bsClass="form-control"
            componentClass="input"
            inputRef={[Function]}
/>,
          <InputGroupButton
            bsClass="input-group-btn"
>
            <Button
                        active={false}
                        block={false}
                        bsClass="btn"
                        bsStyle="default"
                        disabled={false}
                        type="submit"
            >
                        <FormattedMessage
                                    id="predict"
                                    values={Object {}}
                        />
            </Button>
</InputGroupButton>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "form-control",
            "componentClass": "input",
            "inputRef": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "input-group-btn",
            "children": <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              disabled={false}
              type="submit"
>
              <FormattedMessage
                            id="predict"
                            values={Object {}}
              />
</Button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "active": false,
              "block": false,
              "bsClass": "btn",
              "bsStyle": "default",
              "children": <FormattedMessage
                id="predict"
                values={Object {}}
/>,
              "disabled": false,
              "type": "submit",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "id": "predict",
                "values": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bsClass": "form",
        "children": <InputGroup
          bsClass="input-group"
>
          <FormControl
                    bsClass="form-control"
                    componentClass="input"
                    inputRef={[Function]}
          />
          <InputGroupButton
                    bsClass="input-group-btn"
          >
                    <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              disabled={false}
                              type="submit"
                    >
                              <FormattedMessage
                                        id="predict"
                                        values={Object {}}
                              />
                    </Button>
          </InputGroupButton>
</InputGroup>,
        "componentClass": "form",
        "horizontal": false,
        "inline": false,
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsClass": "input-group",
          "children": Array [
            <FormControl
              bsClass="form-control"
              componentClass="input"
              inputRef={[Function]}
/>,
            <InputGroupButton
              bsClass="input-group-btn"
>
              <Button
                            active={false}
                            block={false}
                            bsClass="btn"
                            bsStyle="default"
                            disabled={false}
                            type="submit"
              >
                            <FormattedMessage
                                          id="predict"
                                          values={Object {}}
                            />
              </Button>
</InputGroupButton>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "form-control",
              "componentClass": "input",
              "inputRef": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "input-group-btn",
              "children": <Button
                active={false}
                block={false}
                bsClass="btn"
                bsStyle="default"
                disabled={false}
                type="submit"
>
                <FormattedMessage
                                id="predict"
                                values={Object {}}
                />
</Button>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "active": false,
                "block": false,
                "bsClass": "btn",
                "bsStyle": "default",
                "children": <FormattedMessage
                  id="predict"
                  values={Object {}}
/>,
                "disabled": false,
                "type": "submit",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "id": "predict",
                  "values": Object {},
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`containers/Word Componentが正しく配置されている 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Word
    dispatch={[Function]}
    pos="名詞"
    pronunciation={
        Array [
            Object {
              "id": 1,
              "margin": 100,
              "pronunciation": "やきゅう",
            },
          ]
    }
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
    word="野球"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "active": false,
      "block": false,
      "bsClass": "btn",
      "bsStyle": "default",
      "children": <strong>
        <u>
                野球
        </u>
</strong>,
      "disabled": false,
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <u>
          野球
</u>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "野球",
        },
        "ref": null,
        "rendered": "野球",
        "type": "u",
      },
      "type": "strong",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "active": false,
        "block": false,
        "bsClass": "btn",
        "bsStyle": "default",
        "children": <strong>
          <u>
                    野球
          </u>
</strong>,
        "disabled": false,
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <u>
            野球
</u>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "野球",
          },
          "ref": null,
          "rendered": "野球",
          "type": "u",
        },
        "type": "strong",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
